### ISSUE TITLE
Out-of-bounds read in bcf_unpack_info_core1

### ISSUE THREAD

**Issue by cryptoad** [samtools/htslib#539]
When processed under ASan, the attached file triggers an out-of-bounds read via `view_vcf`:
```
==369612==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6020000024e3 at pc 0x5647659dc310 bp 0x7ffe35258fc0 sp 0x7ffe35258fb8
READ of size 1 at 0x6020000024e3 thread T0
    #0 0x5647659dc30f in bcf_dec_size third_party/htslib/htslib_1_4_1/htslib/vcf.h:993:13
    #1 0x5647659eb65a in bcf_unpack_info_core1 third_party/htslib/htslib_1_4_1/vcf.c:2416:17
    #2 0x5647659eaa34 in bcf_unpack third_party/htslib/htslib_1_4_1/vcf.c:2487:19
    #3 0x5647659ec1ab in vcf_format third_party/htslib/htslib_1_4_1/vcf.c:2504:5
    #4 0x5647659d8129 in vcf_write third_party/htslib/htslib_1_4_1/vcf.c:2615:5
    #5 0x5647659d76ef in bcf_write third_party/htslib/htslib_1_4_1/vcf.c:1514:16
    #6 0x564765959c8b in view_vcf(hFILE*, char const*) experimental/users/kostyak/htsfuzz/htsreader.cc:55:5
    #7 0x5647659595db in main experimental/users/kostyak/htsfuzz/htsreader.cc:85:11

0x6020000024e3 is located 3 bytes to the right of 16-byte region [0x6020000024d0,0x6020000024e0)
allocated by thread T0 here:
    #0 0x564765945ff5 in __interceptor_realloc third_party/llvm/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:79:3
    #1 0x5647659daa5a in ks_resize third_party/htslib/htslib_1_4_1/htslib/kstring.h:113:16
    #2 0x5647659cd0a2 in kputc third_party/htslib/htslib_1_4_1/htslib/kstring.h:161:6
    #3 0x5647659da399 in bcf_enc_size third_party/htslib/htslib_1_4_1/htslib/vcf.h:940:12
    #4 0x5647659da5f0 in bcf_enc_int1 third_party/htslib/htslib_1_4_1/htslib/vcf.h:959:9
    #5 0x5647659df4f1 in vcf_parse third_party/htslib/htslib_1_4_1/vcf.c:2318:21
    #6 0x5647659d0860 in vcf_read third_party/htslib/htslib_1_4_1/vcf.c:2395:12
    #7 0x5647659d0669 in bcf_read third_party/htslib/htslib_1_4_1/vcf.c:1273:42
    #8 0x564765959c5e in view_vcf(hFILE*, char const*) experimental/users/kostyak/htsfuzz/htsreader.cc:54:10
    #9 0x5647659595db in main experimental/users/kostyak/htsfuzz/htsreader.cc:85:11
```
It looks like p just goes past the end of the buffer in `bcf_dec_size`:
```
(gdb) frame 7
#7  0x000055555587a915 in bcf_dec_size (p=0x602000002543 "", q=0x7fffffffc3e0, type=0x60c000001044)
    at third_party/htslib/htslib_1_4_1/htslib/vcf.h:993
993         *type = *p & 0xf;
(gdb) p p 
$1 = (const uint8_t *) 0x602000002543 ""
```
[poc.zip](https://github.com/samtools/htslib/files/1014971/poc.zip)

**Comment by cryptoad** [comment]
Variant with `bcf_dec_typed_int1`: 
```
READ of size 1 at 0x611000000f09 thread T0
    #0 0x565193680cac in bcf_dec_typed_int1 third_party/htslib/htslib_1_4_1/htslib/vcf.h:988:32
    #1 0x5651936785ac in bcf_unpack_info_core1 third_party/htslib/htslib_1_4_1/vcf.c:2415:17
    #2 0x565193677a34 in bcf_unpack third_party/htslib/htslib_1_4_1/vcf.c:2487:19
    #3 0x5651936791ab in vcf_format third_party/htslib/htslib_1_4_1/vcf.c:2504:5
    #4 0x565193665129 in vcf_write third_party/htslib/htslib_1_4_1/vcf.c:2615:5
    #5 0x5651936646ef in bcf_write third_party/htslib/htslib_1_4_1/vcf.c:1514:16
    #6 0x5651935e6c8b in view_vcf(hFILE*, char const*) experimental/users/kostyak/htsfuzz/htsreader.cc:55:5
    #7 0x5651935e65db in main experimental/users/kostyak/htsfuzz/htsreader.cc:85:11

0x611000000f09 is located 9 bytes to the right of 256-byte region [0x611000000e00,0x611000000f00)
allocated by thread T0 here:
    #0 0x5651935d2ff5 in __interceptor_realloc third_party/llvm/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:79:3
    #1 0x565193667a5a in ks_resize third_party/htslib/htslib_1_4_1/htslib/kstring.h:113:16
    #2 0x565193665486 in kputsn third_party/htslib/htslib_1_4_1/htslib/kstring.h:146:6
    #3 0x565193667fa8 in bcf_enc_vchar third_party/htslib/htslib_1_4_1/vcf.c:1809:5
    #4 0x56519366a3fa in vcf_parse third_party/htslib/htslib_1_4_1/vcf.c:2221:25
    #5 0x56519365d860 in vcf_read third_party/htslib/htslib_1_4_1/vcf.c:2395:12
    #6 0x56519365d669 in bcf_read third_party/htslib/htslib_1_4_1/vcf.c:1273:42
    #7 0x5651935e6c5e in view_vcf(hFILE*, char const*) experimental/users/kostyak/htsfuzz/htsreader.cc:54:10
    #8 0x5651935e65db in main experimental/users/kostyak/htsfuzz/htsreader.cc:85:11

```

**Comment by jkbonfield** [comment]
Both this and issue #540 are also fixed by PR #544, so probably come from the same convoluted problem of bad parsing of the VCF headers.

I am unsure at this point if it's worth fixing all the lower level functions (as I did for #537).   We perhaps shouldn't have writing code that relies on the reading code having done sanitisation, but it's a gnarly complex thing.  I'll explore some more though and see how painless the extra write checks are.

**Comment by daviesrob** [comment]
Fixed by #544.  Thanks very much for the report - do let us know if you find anything else.
